cbindings/logic/query.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2025 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cbindings
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/sourcenetwork/defradb/client"
+)
+
+// We cannot return a channel to/from C, so instead we have a map of subscription IDs to
+// RequestResults. Three functions, storeSubscription, getSubscription, and removeSubscription are
+// helpers which manage this store behind the scenes, while PollSubscription and CloseSubscription
+// are made available to the user for interacting with the subscriptions.
+
+var subscriptionStore sync.Map // map[string]*client.RequestResult
+
+type Subscription struct {
+	ctxCancel  context.CancelFunc
+	resultChan <-chan client.GQLResult
+}
+
+// Using UUID lets us avoid collisions, even if we use this across multiple nodes
+func storeSubscription(s Subscription) string {
+	id := uuid.NewString()
+	subscriptionStore.Store(id, s)
+	return id
+}
+
+func getSubscription(id string) (*Subscription, bool) {
+	val, ok := subscriptionStore.Load(id)
+	if !ok {
+		return nil, false
+	}
+	//nolint:forcetypeassert
+	return val.(*Subscription), true
```

Comment:
thought: I'm pretty certain this will panic as the type will be a `Subscription` and not `*Subscription`.

todo: You can either store a pointer or the struct but make sure the asserted type is correct. Note that it doesn't need to be a pointer since both `cancelFunc` and `resultChan` are references.